openfisca_core/commons/formulas.py

```diff
@@ -16,25 +24,41 @@ def apply_thresholds(input, thresholds, choices):
     >>> apply_thresholds(np.array([10]), [5, 7, 9], [10, 15, 20])
     array([0])
     """
+
+    condlist: Sequence[ArrayType[bool]]
+
     condlist = [input <= threshold for threshold in thresholds]
+
     if len(condlist) == len(choices) - 1:
-        # If a choice is provided for input > highest threshold, last condition must be true to return it.
+        # If a choice is provided for input > highest threshold, last condition
+        # must be true to return it.
         condlist += [True]
+
     assert len(condlist) == len(choices), \
-        "apply_thresholds must be called with the same number of thresholds than choices, or one more choice"
+        " ".join([
+            "apply_thresholds must be called with the same number of",
+            "thresholds than choices, or one more choice",
+            ])
+
     return numpy.select(condlist, choices)
 
 
-def concat(this, that):
-    if isinstance(this, numpy.ndarray) and not numpy.issubdtype(this.dtype, numpy.str):
+def concat(this: ArrayLike[str], that: ArrayLike[str]) -> ArrayType[str]:
+    if isinstance(this, numpy.ndarray) and \
+       not numpy.issubdtype(this.dtype, numpy.str_):
         this = this.astype('str')
-    if isinstance(that, numpy.ndarray) and not numpy.issubdtype(that.dtype, numpy.str):
+
+    if isinstance(that, numpy.ndarray) and \
+       not numpy.issubdtype(that.dtype, numpy.str_):
         that = that.astype('str')
 
-    return numpy.core.defchararray.add(this, that)
+    return numpy.char.add(this, that)
 
 
-def switch(conditions, value_by_condition):
+def switch(
+        conditions: ArrayType[float],
```

Comment:
The conditions can be of any type.

Reply:
Any type of array?

Reply:
Yes boolean, int, enums etc. (all those might be included in float and thus duck typing might work).

Reply:
Ok thanks for the clarification, I wrote a test for this an you're right, typing is wrong 🙏  …